nav2_smac_planner/src/node_hybrid.cpp

```diff
@@ -492,6 +515,126 @@ float NodeHybrid::getObstacleHeuristic(
+        goal_index, size_x, start_x,
+        start_y), goal_index);
+
+    // initialize goal cell with a very small value to differentiate it from 0.0 (~uninitialized)
+    // the negative value means the cell is in the open set
+    obstacle_heuristic_lookup_table[goal_index] = -0.00001f;
+  } else {
+    for (auto & n : astar_2d_queue) {
+      n.first = -obstacle_heuristic_lookup_table[n.second] +
+        distanceHeuristic2D(n.second, size_x, start_x, start_y);
+    }
+    std::make_heap(astar_2d_queue.begin(), astar_2d_queue.end(), AStar2DComparator{});
+  }
+
+  float c_cost, cost, travel_cost, new_cost, existing_cost;
+  int mx, my, mx_idx, my_idx;
+  int new_idx = 0;
+  const float sqrt_2 = sqrt(2);
+
+  while (!astar_2d_queue.empty()) {
+    AStar2DElement c = astar_2d_queue.front();
+    std::pop_heap(astar_2d_queue.begin(), astar_2d_queue.end(), AStar2DComparator{});
+    astar_2d_queue.pop_back();
+    c_cost = obstacle_heuristic_lookup_table[c.second];
+    if (c_cost > 0.0f) {
+      // cell has been processed and closed, no further cost improvements
+      // are mathematically possible thanks to euclidean heuristic consistency
+      continue;
+    }
+    c_cost = -c_cost;
+    obstacle_heuristic_lookup_table[c.second] = c_cost;  // set a positive value to close the cell
+
+    my_idx = c.second / size_x;
+    mx_idx = c.second - (my_idx * size_x);
+
+    // find neighbors
+    for (unsigned int i = 0; i != neighborhood.size(); i++) {
+      new_idx = static_cast<unsigned int>(static_cast<int>(c.second) + neighborhood[i]);
+      cost = static_cast<float>(sampled_costmap->getCost(new_idx));
+
+      // if neighbor path is better and non-lethal, set new cost and add to queue
+      if (new_idx > 0 && new_idx < size_x * size_y && cost < INSCRIBED) {
+        my = new_idx / size_x;
+        mx = new_idx - (my * size_x);
+
+        if (mx == 0 && mx_idx >= size_x - 1 || mx >= size_x - 1 && mx_idx == 0) {
+          continue;
+        }
+        if (my == 0 && my_idx >= size_y - 1 || my >= size_y - 1 && my_idx == 0) {
+          continue;
+        }
+
+        existing_cost = obstacle_heuristic_lookup_table[new_idx];
+        if (existing_cost <= 0.0f) {
+          travel_cost =
+            ((i <= 3) ? 1.0f : sqrt_2) * (1.0f + (cost_penalty * cost / 252.0f));
+          new_cost = c_cost + travel_cost;
+          if (existing_cost == 0.0f || -existing_cost > new_cost) {
+            // the negative value means the cell is in the open set
+            obstacle_heuristic_lookup_table[new_idx] = -new_cost;
+            astar_2d_queue.emplace_back(
+              new_cost +
+              distanceHeuristic2D(new_idx, size_x, start_x, start_y), new_idx);
+            std::push_heap(astar_2d_queue.begin(), astar_2d_queue.end(), AStar2DComparator{});
+          }
+        }
+      }
+    }
+
+    if (c.second == start_index) {
+      break;
+    }
+  }
+
+  // costs are doubled due to downsampling
+  return 2.0 * starting_cost;
```

Comment:
The fact that `starting_cost` is a reference is a little subtle. Mind changing this to be a little more explicit `obstacle_heuristic_lookup_table[start_index];` or adding a little comment here that `starting_cost` is a reference so its updated from the search process (I don't mind which)?

Reply:
Will add a comment and rename `starting_cost` to `requested_cell_cost`